src/NServiceBus.Persistence.DynamoDB/Outbox/OutboxPersister.cs

```diff
@@ -36,66 +37,96 @@ public OutboxPersister(IAmazonDynamoDB dynamoDbClient, OutboxPersistenceConfigur
-                allItems.AddRange(response.Items);
+                bool headerItemSet = false;
+                if (incomingId == null && response.Items.Count >= 1)
+                {
+                    var headerItem = response.Items[0];
+                    incomingId = headerItem[configuration.PartitionKeyName].S;
+                    numberOfTransportOperations = Convert.ToInt32(headerItem["TransportOperations"].N);
+                    headerItemSet = true;
+                }
+
+                for (int i = headerItemSet ? 1 : 0; i < response.Items.Count; i++)
+                {
+                    transportOperationsAttributes ??= new List<Dictionary<string, AttributeValue>>(numberOfTransportOperations);
+                    transportOperationsAttributes.Add(response.Items[i]);
+                }
             } while (response.LastEvaluatedKey.Count > 0);
 
-            if (allItems.Count == 0)
-            {
+            return incomingId == null ?
                 //TODO: Should we check the response code to throw if there is an error (other than 404)
-                return null;
-            }
-
-            return DeserializeOutboxMessage(allItems, context);
+                null : DeserializeOutboxMessage(incomingId, numberOfTransportOperations, transportOperationsAttributes, context);
         }
 
-        OutboxMessage DeserializeOutboxMessage(List<Dictionary<string, AttributeValue>> responseItems,
-            ContextBag contextBag)
+        OutboxMessage DeserializeOutboxMessage(string incomingId, int numberOfTransportOperations,
+            List<Dictionary<string, AttributeValue>> transportOperationsAttributes, ContextBag contextBag)
         {
-            var headerItem = responseItems.First();
-            var incomingId = headerItem[configuration.PartitionKeyName].S;
-            // TODO: In case we delete stuff we can probably even remove this property
-            int numberOfTransportOperations = Convert.ToInt32(headerItem["TransportOperations"].N);
+            // Using numberOfTransportOperations instead of transportOperationsAttributes.Count to account for
+            // potential partial deletes
             contextBag.Set(OperationsCountContextProperty, numberOfTransportOperations);
 
-            var operations = Array.Empty<TransportOperation>();
-            if (numberOfTransportOperations > 0)
+            var operations = numberOfTransportOperations == 0
+                ? Array.Empty<TransportOperation>()
+                : new TransportOperation[numberOfTransportOperations];
+
+            for (int i = 0; i < numberOfTransportOperations; i++)
             {
-                operations = responseItems.Skip(1).Select(DeserializeOperation).ToArray();
+                operations[i] = DeserializeOperation(transportOperationsAttributes![i]);
             }
 
             return new OutboxMessage(incomingId, operations);
```

Comment:
is that even correct? This isn't really the value of the incoming message id as the property states

Reply:
Yeah that was not correct in the original code. Good catch